torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py

```diff
@@ -0,0 +1,165 @@
+    """
+
+    def __init__(self, period_group_size_dict=None, warmup_steps=0, process_group=None):
+        if not period_group_size_dict:
+            raise ValueError("Arg ``period_group_size_dict`` must not be empty.")
+        self._periods = list(period_group_size_dict.keys())
+        if self._periods[0] < 0:
+            raise ValueError("The minimum period in arg ``period_group_size_dict`` must be a positive value.")
+        elif self._periods[-1] == 1:
+            warnings.warn(
+                "When the maximum period in arg ``period_group_size_dict`` is 1, "
+                "no need to use model averaging because the communication cost "
+                "of all-reducing parameters will be no less than the cost of all-reducing gradients "
+                "by DistributedDataParall in the backward pass. Therefore, only "
+                "DistributedDataParallel should be used for this case."
+            )
+        ovall_group : dist.ProcessGroup = (
+            process_group if process_group is not None else dist.group.WORLD
+        )
+        overall_group_size = dist.get_world_size(group=ovall_group)
+        if list(period_group_size_dict.values())[-1] != overall_group_size:
+            raise ValueError(
+                "The last value in arg ``period_process_group_dict`` "
+                "must be equal to the size of arg ``process_group``.")
+
+        self.period_process_group_dict = OrderedDict()
+        logger.info("Model averaging hierarchy:")
+        for period, group_size in period_group_size_dict.items():
+            logger.info(
+                f"\tEach group that has {group_size} processes average parameters every {period} iterations, "
+                "if no higher-level averaging.")
+            if group_size != overall_group_size:
+                self.period_process_group_dict[period], subgroups = dist.new_subgroups(
+                    group_size=group_size, group=ovall_group)
+            else:
+                self.period_process_group_dict[period] = ovall_group
+
+        if warmup_steps < 0:
+            raise ValueError("Arg ``warmup_steps`` must be a non-negative number.")
+        self.warmup_steps = warmup_steps
+        self.step = 0
+        self.group = ovall_group  # The process group used for averaging parameters.
+
+    def _is_process_group_found(self):
+        """
+        Returns whether ``step`` can be divided by a period in the keys of ``period_process_group_dict``.
+        If ``step`` can be divided by multiple periods in the keys of ``period_process_group_dict``,
+        then sets set ``_group`` as the process group corresponding to the largest period,
+        since this process group will be used for averaging parameters at this ``step``.
+        """
+        for period in reversed(self._periods):
+            if self.step % period == 0:
+                self.group = self.period_process_group_dict[period]
+                return True
+        return False
```

Comment:
not the biggest fan of a function modifying state unless absolutely necessary, but fine with this as is.

One suggestion is that this function can return `Tuple[bool, Optional[ProcessGroup]]` where the latter is `None` if this function is returning `False`, otherwise it contains the process group to use. 

this will also help get rid of `self.group` entirely?

Reply:
Agree with your proposed change if we want to get rid of `self.group`. Actually I prefer keeping this state for now, as it can help me with some debugging. For example, even at a step that does not run model averaging, I still want to know what's the group size of the most recent model averaging.